server/handler/http_funtional_cluster_test.go

```diff
@@ -0,0 +1,737 @@
+			AllowedStatuses: []int{200, 301, 302},
+			AllowedMethods:  []string{"HEAD", "GET"},
+		},
+		CircuitBreaker: circuit_breaker.CircuitBreaker{
+			Threshold:   2,                // after 2nd request, if meet FailureRate goes open.
+			FailureRate: 0.5,              // 1 out of 2 fails, or more
+			Interval:    time.Duration(1), // clears counts immediately
+			Timeout:     time.Duration(1), // clears state immediately
+		},
+	}
+
+	domainID := config.Config.Server.Upstream.GetDomainID()
+	balancer.InitRoundRobin(domainID, config.Config.Server.Upstream, false)
+	circuit_breaker.InitCircuitBreaker(domainID, config.Config.CircuitBreaker, logger.GetGlobal())
+	engine.InitConn(domainID, config.Config.Cache, log.StandardLogger())
+
+	_, _ = engine.GetConn(domainID).PurgeAll()
+
+	// --- MISS
+
+	req, err := http.NewRequest("GET", "/standards/", nil)
+	req.URL.Scheme = config.Config.Server.Upstream.Scheme
+	req.URL.Host = config.Config.Server.Upstream.Host
+	req.Host = config.Config.Server.Upstream.Host
+	req.TLS = &tls.ConnectionState{} // mock a fake https
+	assert.Nil(t, err)
+
+	rr := httptest.NewRecorder()
+	h := http.HandlerFunc(handler.HandleRequest)
+
+	h.ServeHTTP(rr, req)
+
+	assert.Equal(t, http.StatusOK, rr.Code)
+
+	assert.Equal(t, "MISS", rr.HeaderMap["X-Go-Proxy-Cache-Status"][0])
+
+	body := rr.Body.String()
+
+	assert.Contains(t, body, "<!doctype html>")
+	assert.Contains(t, body, "<title>W3C</title>")
+	assert.Contains(t, body, "</body>\n\n</html>\n")
+
+	// --- HIT
+
+	req, err = http.NewRequest("GET", "/standards/", nil)
+	req.URL.Scheme = config.Config.Server.Upstream.Scheme
+	req.URL.Host = config.Config.Server.Upstream.Host
+	req.Host = config.Config.Server.Upstream.Host
+	req.TLS = &tls.ConnectionState{} // mock a fake https
+	assert.Nil(t, err)
+
+	rr = httptest.NewRecorder()
+	h.ServeHTTP(rr, req)
+
+	assert.Equal(t, http.StatusOK, rr.Code)
+
+	assert.Equal(t, "HIT", rr.HeaderMap["X-Go-Proxy-Cache-Status"][0])
+
+	body = rr.Body.String()
+
+	assert.Contains(t, body, "<!doctype html>")
+	assert.Contains(t, body, "<title>W3C</title>")
+	assert.Contains(t, body, "</body>\n\n</html>\n")
+
+	// Manual Timeout All Fresh Keys
+	_, _ = engine.GetConn(domainID).DelWildcard(context.Background(), "DATA@@GET@@https://www.w3.org/standards/@@*/fresh")
+
+	// --- STALE
+
+	req, err = http.NewRequest("GET", "/standards/", nil)
+	req.URL.Scheme = config.Config.Server.Upstream.Scheme
+	req.URL.Host = config.Config.Server.Upstream.Host
+	req.Host = config.Config.Server.Upstream.Host
+	req.TLS = &tls.ConnectionState{} // mock a fake https
+	assert.Nil(t, err)
+
+	rr = httptest.NewRecorder()
+	h.ServeHTTP(rr, req)
+
+	assert.Equal(t, http.StatusOK, rr.Code)
+
+	assert.Equal(t, "STALE", rr.HeaderMap["X-Go-Proxy-Cache-Status"][0])
+
+	body = rr.Body.String()
+
+	assert.Contains(t, body, "<!doctype html>")
+	assert.Contains(t, body, "<title>W3C</title>")
+	assert.Contains(t, body, "</body>\n\n</html>\n")
+
+	tearDownHTTPFunctionalCluster()
+}
+
```

Comment:
By the look of it, each test file is replicated for the cluster functionality. There's one test case missing in this file `TestHTTPEndToEndCallWithHTTPSRedirect`.

Reply:
In this case there is no cluster test because there is no redis host set in the original test, so it doesn´t have impact on it.

Reply:
ok, got it, makes sense.